### PR TITLE
configure wildfly heap based on container memory

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@gabemontero ptal.  our wildfly image was horribly out of date w/ respect to handling jvm heap configuration.

File: 10.0/contrib/wfbin/standalone.conf [hunk 1]
```diff
@@ -91,9 +86,25 @@ if [ -z "$OPENSHIFT_SMTP_HOST" ]; then
     export OPENSHIFT_SMTP_HOST="localhost"
 fi
 
-max_memory_mb=${OPENSHIFT_GEAR_MEMORY_MB} 
-if [ -z "$OPENSHIFT_GEAR_MEMORY_MB" ]; then
-  max_memory_mb=512
+if [ -z $JAVA_GC_OPTS ]; then
+  JAVA_GC_OPTS="-XX:+UseParallelGC -XX:MaxPermSize=100m -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=40 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90"
```
Comment (gabemontero) [inline 10.0/contrib/wfbin/standalone.conf]: Can the user vary jdk7 vs. jdk8 with this image?  If only jdk8, MaxPermSize is ignored, and a benign by noisy info log shows up.

File: 10.0/contrib/wfbin/standalone.conf [hunk 2]
```diff
@@ -103,9 +114,9 @@ messaging_scheduled_thread_pool_max_size=5
 
 if [ $max_heap -lt 1024 ]
 then
-    memory_options="-XX:+UseParallelGC -Xms40m -Xmx${max_heap}m -XX:+AggressiveOpts -XX:MaxPermSize=100m -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=40 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dorg.apache.tomcat.util.LOW_MEMORY=true" 
+    memory_options="-XX:+UseParallelGC -Xms40m -Xmx${max_heap}m -XX:+AggressiveOpts -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=40 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dorg.apache.tomcat.util.LOW_MEMORY=true"
```
Comment (gabemontero) [inline 10.0/contrib/wfbin/standalone.conf]: Ah - I see you removed MaxPermSize here

File: 10.1/contrib/wfbin/standalone.conf [hunk 3]
```diff
@@ -91,9 +86,25 @@ if [ -z "$OPENSHIFT_SMTP_HOST" ]; then
     export OPENSHIFT_SMTP_HOST="localhost"
 fi
 
-max_memory_mb=${OPENSHIFT_GEAR_MEMORY_MB} 
-if [ -z "$OPENSHIFT_GEAR_MEMORY_MB" ]; then
-  max_memory_mb=512
+if [ -z $JAVA_GC_OPTS ]; then
+  JAVA_GC_OPTS="-XX:+UseParallelGC -XX:MaxPermSize=100m -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=40 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90"
```
Comment (gabemontero) [inline 10.1/contrib/wfbin/standalone.conf]: same perm size comment here

File: 10.0/contrib/wfbin/standalone.conf [hunk 4]
```diff
@@ -91,9 +86,25 @@ if [ -z "$OPENSHIFT_SMTP_HOST" ]; then
...
+# otherwise try to set the container memory size based on the cgroup limits
+else
+  CONTAINER_MEMORY_IN_BYTES=`cat /sys/fs/cgroup/memory/memory.limit_in_bytes`
+  DEFAULT_MEMORY_CEILING=$((2**40-1))
+  if [ "${CONTAINER_MEMORY_IN_BYTES}" -lt "${DEFAULT_MEMORY_CEILING}" ]; then
+      max_memory_mb=$((${CONTAINER_MEMORY_IN_BYTES}/1024**2))
+  fi 
+fi
+if [ -z "$JVM_HEAP_RATIO" ]; then
+    JVM_HEAP_RATIO=0.5
```
Comment (gabemontero) [inline 10.0/contrib/wfbin/standalone.conf]: Assuming 64 bit only, are we concerned with the -Xss setting we do (until the 32bit change go in) in the jenkins image?

File: 10.1/contrib/wfbin/standalone.conf [hunk 5]
```diff
@@ -91,9 +86,25 @@ if [ -z "$OPENSHIFT_SMTP_HOST" ]; then
...
+# otherwise try to set the container memory size based on the cgroup limits
+else
+  CONTAINER_MEMORY_IN_BYTES=`cat /sys/fs/cgroup/memory/memory.limit_in_bytes`
+  DEFAULT_MEMORY_CEILING=$((2**40-1))
+  if [ "${CONTAINER_MEMORY_IN_BYTES}" -lt "${DEFAULT_MEMORY_CEILING}" ]; then
+      max_memory_mb=$((${CONTAINER_MEMORY_IN_BYTES}/1024**2))
+  fi 
+fi
+if [ -z "$JVM_HEAP_RATIO" ]; then
+    JVM_HEAP_RATIO=0.5
```
Comment (gabemontero) [inline 10.1/contrib/wfbin/standalone.conf]: Same -Xss comment here

File: 10.0/contrib/wfbin/standalone.conf (hunk 1 above)
Comment (bparees) [inline 10.0/contrib/wfbin/standalone.conf]: this env actually shouldn't be here at all, copy+paste+edit fail.

File: 10.0/contrib/wfbin/standalone.conf (hunk 4 above)
Comment (bparees) [inline 10.0/contrib/wfbin/standalone.conf]: no, I don't think wildfly has the same memory problems that jenkins does, at least not that i've witnessed so far, so i chose to leave that complexity out, at least for now.

**Comment by bparees** [comment]
@gabemontero updated

**Comment by gabemontero** [comment]
IGTM

**Pull request merged**